cirq/circuits/text_diagram_drawer.py

```diff
@@ -221,20 +253,74 @@ def _cross_char(use_ascii: bool, horizontal_emph: bool, vertical_emph: bool
+                length: int,
+                spacing: int,
+                moment_groups: List[_MomentGroup],
+                moment_group_parts: _MomentGroupParts) -> int:
+    """Pad a border with indicators for Moment groups.
+
+    Args:
+        border: A list of cells corresponding to the border of a circuit grid.
+        index: The index of the border cell to be padded.
+        length: The width (if horizontal) or height (if vertical) of the
+            longest contents of a cell.
+        spacing: The number of spaces (if horizontal) or rows (if vertical)
+            between this border cell and the next. If spacing is 2 or greater,
+            the start and end of each interval is drawn in the adjacent cell.
+        moment_groups: A set of intervals each of which groups together
+            columns (if horizontal) or rows (if vertical) belonging to
+            the same Moment.
+        moment_group_parts: The characters to use for drawing the start,
+            middle, and end of an interval indicating a Moment group.
+
+    Returns:
+        The new length for that border cell.
+    """
+    (start_char, middle_char, end_char) = moment_group_parts
+    overshoot = spacing >= 2
+    if any(index in range(start, end) for (start, end) in moment_groups):
+        border.append(middle_char * length)
+    else:
+        new_length = length + spacing
+        if any(index == end for (_, end) in moment_groups):
+            middle_char_length = length if overshoot else length - 1
+            border.append((middle_char * middle_char_length +
+                           end_char).ljust(new_length, ' '))
+        else:
+            border.append(' ' * new_length)
+        length = new_length
+    if any(index == start for (start, _) in moment_groups):
+        if overshoot:
+            border[index - 1] = border[index - 1][:-1] + start_char
+        else:
+            border[index] = start_char + border[index][1:]
+    return length
+
+
 def _pad_into_multiline(width: int,
                         grid: List[List[str]],
                         horizontal_separator: List[List[str]],
                         vertical_separator: List[List[str]],
                         horizontal_spacing: int,
-                        vertical_spacing: int
+                        vertical_spacing: int,
+                        moment_groups: List[_MomentGroup],
+                        moment_group_parts: _MomentGroupParts,
+                        groups_are_vertical: bool
                         ) -> List[List[List[str]]]:
     multiline_grid = []  # type: List[List[List[str]]]
+    border = []
 
     # Vertical padding.
     for row in range(len(grid)):
         multiline_cells = [cell.split('\n') for cell in grid[row]]
         row_height = max(1, max(len(cell) for cell in multiline_cells))
-        row_height += vertical_spacing
+        if groups_are_vertical:
+            row_height = _pad_border(border, row, row_height, vertical_spacing,
```

Comment:
This makes the previous `row_height` assignment a dead write in this case. Move that previous assignment into the `if`'s positive case.

Reply:
Oops. Obsoleted by my huge design change comment.